common/src/Model/TagManager.h

```diff
@@ -0,0 +1,93 @@
+/*
+ Copyright (C) 2010-2017 Kristian Duske
+
+ This file is part of TrenchBroom.
+
+ TrenchBroom is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ TrenchBroom is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with TrenchBroom. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef TRENCHBROOM_TAGMANAGER_H
+#define TRENCHBROOM_TAGMANAGER_H
+
+#include "StringUtils.h"
+#include "Model/Tag.h"
+
+#include <vector>
+
+namespace TrenchBroom {
+    namespace Model {
+        /**
+         * Manages the tags used in a document and updates smart tags on taggable objects.
+         */
+        class TagManager {
+        private:
+            size_t m_currentTagTypeIndex;
+            std::vector<SmartTag> m_smartTags;
```

Comment:
Adding or removing from this vector will break any of the TagReferences in Taggable objects, since the vector might resize. I've caused this bug before in other codebases and it's really annoying.

Reply:
Yeah, I'll switch this to a list. Or I could store unique_ptr in the vector.